### PR TITLE
Refactor token authentication plugin to use new luajwtjitsi version

### PR DESCRIPTION
**Pull request by philip-cc**
This PR updates the Prosody token authentication plugin to use luajwtjitsi 3.0, mitigating the risk relating to [this](https://github.com/jitsi/jitsi-meet/security/advisories/GHSA-45ff-37jm-xjfx) recent jitsi-meet security advisory. See also the related PR: https://github.com/jitsi/luajwtjitsi/pull/3.

The main changes are:
* Update luajwtjitsi from version 2.0 to 3.0
* Removed `Util:verify_issuer`, `Util:verify_audience`, `Util:verify_token` as these are now part of luajwtjitsi. Note that the check for `room` claims is preserved.
* Added a `signatureAlgorithm` option to follow best practices of the server dictating the signature algorithm for use. 
* Slight refactor in how a symmetric signing key is set.

### CODE REVIEW

**Comment by saghul** [comment]
I pushed 3.0 to luarocks, BTW.

**Comment by damencho** [comment]
The new version will be installed just for new deployments, right?
If that is correct, all that are already using it and update will be in a broken state.

File: resources/prosody-plugins/token/util.lib.lua [hunk 1]
```diff
@@ -33,6 +33,7 @@ function Util.new(module)
     self.appId = module:get_option_string("app_id");
     self.appSecret = module:get_option_string("app_secret");
     self.asapKeyServer = module:get_option_string("asap_key_server");
+    self.signatureAlgorithm = module:get_option_string("signature_algorithm");
```
Comment (damencho) [inline resources/prosody-plugins/token/util.lib.lua:36]: Now there is an extra configuration that we depend on, without default value and all the users that update will be broken till they update their config? 
I also do not see this option added to the default prosody template that is used for new installs.
Why is this needed to be configurable? Isn't it that when you use a shared secret (module:get_option_string("app_secret");) it is HS256 and when using "asap_key_server" it is RS256?

Comment (philip-cc) [reply resources/prosody-plugins/token/util.lib.lua:36]: > Isn't it that when you use a shared secret (module:get_option_string("app_secret");) it is HS256 and when using "asap_key_server" it is RS256?
luajwtjitsi supports larger key sizes (e.g. HS512/RS512) which I would definitely recommend you to support. But it is probably good to have it default to HS256/RS256 when `app_secret`/`asap_key_server` is set, and then allow to overwrite when `signature_algorithm` is set. I will update with that change.

Comment (philip-cc) [reply resources/prosody-plugins/token/util.lib.lua:36]: I have added defaults for the `signature_algorithm` option based on the user of `app_secret`/`asap_key_server`. That should preserve the current functionality.

**Comment by philip-cc** [comment]
Hi, is there anything else I need to update on this one?

File: resources/prosody-plugins/token/util.lib.lua [hunk 2]
```diff
@@ -274,23 +198,38 @@ function Util:process_and_verify_token(session, acceptedIssuers)
...
-    else
-        claims, msg = self:verify_token(session.auth_token, self.appSecret, acceptedIssuers);
+    if key == nil then
+        return false, "not-allowed", "signature verification key is missing";
     end
+
+    -- now verify the whole token
+    local claims, msg = jwt.verify(
+        session.auth_token,
+        signatureAlgorithm,
```
Comment (saghul) [inline resources/prosody-plugins/token/util.lib.lua:220]:
```suggestion
        self.signatureAlgorithm,
```

**Review by saghul — APPROVED** [review]

**Comment by saghul** [comment]
Jenkins please test this please.

**Pull request merged**